### PR TITLE
drivers: crypto: Add NXP MCUX DCP driver

### PR DESCRIPTION
**Pull request by pdgendt**
This PR enables the Data Co-Processor present in NXP i.MX RT10XX SoCs. It also simplifies the sample to use a common `prj.conf`.

Note: I did not get the driver to work with DCACHE enabled, hence the driver disables/enables this when calling the HW peripheral.

### CODE REVIEW

**Comment by mmahadevan108** [comment]
@pdgendt The HAL PR has been merged. Can you please update `west.yml`

**Comment by mmahadevan108** [comment]
@pdgendt Kindly help resolve the conflicts. The changes to `west.yml` is no longer needed.

**Comment by pdgendt** [comment]
@mmahadevan108 I've rebased on `main`

I'm not particularly happy with the `sys_cache_data_enable/_disable` workarounds, if someone has suggestions to improve this, I'd appreciate it.

**Review by danieldegrasse — APPROVED** [review]
DCP driver looks good to me, although I am not an expert on crypto APIs. @ceolin, could you look at the changes to the crypto driver sample?

> I'm not particularly happy with the sys_cache_data_enable/_disable workarounds, if someone has suggestions to improve this, I'd appreciate it.

From what I can tell, the input and output buffers for the DCP engine either need to be in noncacheable memory, or aligned to the cache line size (32 bytes). I was able to get the `crypto_dcp_aes_ecb_encrypt` with `sys_cache_data_flush_range` and `sys_cache_data_invd_range` by using an intermediate set of buffers aligned to 32 byte boundaries, and passing those to the DCP.

Given that there is no way at the API level to set this alignment requirement, I don't think I'm aware of a better approach for this right now.

**Comment by ceolin** [comment]
> @mmahadevan108 I've rebased on `main`
> 
> I'm not particularly happy with the `sys_cache_data_enable/_disable` workarounds, if someone has suggestions to improve this, I'd appreciate it.

The buffer is given by the caller so you don't really know if it is a cachable/non-cachable memory and the packet buffers are not allocated by shim so we can't enforce alignment. I agree, that is probably the best solution for now.

**Review by ceolin — APPROVED** [review]

File: drivers/crypto/crypto_mcux_dcp.c [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+	return 0;
+}
+
+static int crypto_dcp_cipher_free_session(const struct device *dev, struct cipher_ctx *ctx)
+{
+	struct crypto_dcp_session *session;
+
+	ARG_UNUSED(dev);
+
+	session = ctx->drv_sessn_state;
```
[the remark this answers is not in the document]
Comment (pdgendt) [reply drivers/crypto/crypto_mcux_dcp.c:240]: This function is used in the `cipher` part of the API, not for hashing?

**Pull request merged**